beps/0001-notifications-system/README.md

```diff
@@ -147,6 +148,8 @@ The link is a relative or absolute URL. As an example, it can be used:
 - by an external system to request an action within an asynchronous task
 - by a BE plugin to provide link to other part of the Backstage UI (i.e. to the Catalog)
 
+The extra is a flexible JSON like field, where an additional payload can be stored.
```

Comment:
Yep I think we need this. It would be good to have a bit more clarity in who produces and consumes this though. Who owns the format of the extra fields and who is allowed to look at the data.

For example, fairly straightforward I think would be that a plugin that posts a notification can then use the extra information for customized rendering in the frontend. But are notification processors allowed to access that data for custom logic, and what are the stability guarantees of the data in that case?

Reply:
We would like to build a template, where stakeholders from technical platform teams could notify all affected feature teams what has been changed and which applications were affected, so teams have to take an action, they have a deadline for that, etc. Once the action by team has been done they mark it as completed. This way we can get rid of using emails, which sometimes include also more teams then needed.

Most of the fields which are currently in "notification" table perfectly fit this purpose, but some fields are missing (of course 😄 ) which we would like to store.

In some mid-term we would like to add some other notifications which comes from security point of view, and the template, number of fields are different from the first template. 

Therefore having a "notification" for our case should be flexible to support it. 

Maybe we can have something like this:

```
{
  "data": {
    "deadline": "01-06-2024",
    "affectedTeams": [
      "team1",
      "team2"
    ]
  },
  "schema": {
    "module": "central-hub-notification-module",
    "fields": [
      {
        "name": "deadline",
        "format": date,
        "optional": false
      },
      {
        "affectedTeams": string[]
      }
    ]
  }
}
```

This can be an indication which module can process an 'extra' field. 

If to think more globally, even one more field in the "Notification" or "NotificationPayload" object could be added, **module owner/source owner**, and only that module can retrieve this notification. 

So it's a matter of limiting the access only to "extra" field, or completely filtering out notifications based on custom notification module

Reply:
How about adding a `notification.metadata` as `JsonValue | undefined` to the notification payload and saving that as JSON to the database? Also, I think this data should be available for the processors to modify/read as needed. The default processors (email/slack/...) from upstream probably don't know what to do with this data and should not care but custom processors can do their things with the data.

Reply:
Yeah, absolutely as part of the `notificationPayload`.